tensorboard/defs/internal/html.bzl

```diff
@@ -0,0 +1,133 @@
+    for dep in deps:
+        manifests = depset(transitive = [manifests, dep.webfiles.manifests])
+        webpaths = depset(transitive = [webpaths, dep.webfiles.webpaths])
+        files = depset(transitive = [files, dep.data_runfiles.files])
+    webpaths = depset([ctx.attr.output_path], transitive = [webpaths])
+    closure_js_library = collect_js(
+        unfurl(ctx.attr.deps, provider = "closure_js_library"),
+    )
+
+    # vulcanize
+    ignore_regexs_file_set = depset()
+    ignore_regexs_file_path = "NO_REGEXS"
+    ctx.actions.run(
+        inputs = depset(transitive = [
+            manifests,
+            files,
+        ]).to_list(),
+        outputs = [ctx.outputs.html, ctx.outputs.js],
+        executable = ctx.executable._Vulcanize,
+        arguments = ([
+                         ctx.attr.input_path,
+                         ctx.attr.output_path,
+                         ctx.attr.js_path,
+                         ctx.outputs.html.path,
+                         ctx.outputs.js.path,
+                     ] +
+                     [f.path for f in manifests.to_list()]),
+        mnemonic = "Vulcanize",
+        progress_message = "Vulcanizing %s" % ctx.attr.input_path,
+    )
+
+    # webfiles manifest
+    manifest_srcs = [struct(
+        path = ctx.outputs.html.path,
+        longpath = long_path(ctx, ctx.outputs.html),
+        webpath = ctx.attr.output_path,
+    )]
+
+    if ctx.attr.js_path:
+        manifest_srcs.append(
+            struct(
+                path = ctx.outputs.js.path,
+                longpath = long_path(ctx, ctx.outputs.js),
+                webpath = ctx.attr.js_path,
+            ),
+        )
+
+    manifest = ctx.actions.declare_file("%s.pbtxt" % ctx.label.name)
+    ctx.actions.write(
+        output = manifest,
+        content = struct(
+            label = str(ctx.label),
+            src = manifest_srcs,
+        ).to_proto(),
+    )
+    manifests = depset([manifest], transitive = [manifests])
+
+    transitive_runfiles = depset()
+    for dep in deps:
+        transitive_runfiles = depset(transitive = [
+            transitive_runfiles,
+            dep.data_runfiles.files,
+        ])
+
+    return struct(
+        files = depset([ctx.outputs.html, ctx.outputs.js]),
+        webfiles = struct(
+            manifest = manifest,
+            manifests = manifests,
+            webpaths = webpaths,
+            dummy = ctx.outputs.html,
+        ),
+        runfiles = ctx.runfiles(
+            files = ctx.files.data + [
+                manifest,
+                ctx.outputs.html,
+                ctx.outputs.js,
+            ],
+            transitive_files = transitive_runfiles,
+        ),
+    )
+
+tb_combine_html = rule(
+    implementation = _tb_combine_html_impl,
+    attrs = {
+        "input_path": attr.string(mandatory = True),
+        "output_path": attr.string(mandatory = True),
+        # If specified, it extracts scripts into {name}.js and inserts <script src="{js_path}">.
```

Comment:
Could we use `doc` parameters for these attrs to provide a brief description of each one, just for posterity so we can better understand how this is used?  They can just be a single line.

One thing that I had to go read `Vulcanize.java` to remember is the distinction between `input_path` and `deps` - i.e. that `input_path` is where we start the crawl from, and `deps` has to contain all the webfiles that `input_path` is referencing.  Just brief mentions of things like that would help contextualize what's happening here.

Reply:
I gave it a stab. Thanks for the suggestion.